common/persistence/query_util.go

```diff
@@ -57,22 +66,106 @@ func LoadAndSplitQuery(
 func LoadAndSplitQueryFromReaders(
 	readers []io.Reader,
 ) ([]string, error) {
-
 	result := make([]string, 0, querySliceDefaultSize)
-
 	for _, r := range readers {
 		content, err := io.ReadAll(r)
 		if err != nil {
 			return nil, fmt.Errorf("error reading contents: %w", err)
 		}
-		for _, stmt := range strings.Split(string(content), queryDelimiter) {
-			stmt = strings.TrimSpace(stmt)
+		n := len(content)
+		contentStr := string(bytes.ToLower(content))
+		for i, j := 0, 0; i < n; i = j {
+			// stack to keep track of open parenthesis/blocks
+			var st []byte
+
+		stmtLoop:
+			for ; j < n; j++ {
+				switch contentStr[j] {
+				case queryDelimiter:
+					if len(st) == 0 {
+						j += 1
+						break stmtLoop
+					}
+
+				case sqlLeftParenthesis:
+					st = append(st, sqlLeftParenthesis)
+
+				case sqlRightParenthesis:
+					if len(st) == 0 || st[len(st)-1] != sqlLeftParenthesis {
+						return nil, fmt.Errorf("error reading contents: unmatched right parenthesis")
+					}
+					st = st[:len(st)-1]
+
+				case sqlBeginKeyword[0]:
+					if matchWord(contentStr, sqlBeginKeyword, j) {
+						st = append(st, sqlBeginKeyword[0])
+						j += len(sqlBeginKeyword) - 1
+					}
+
+				case sqlEndKeyword[0]:
+					if matchWord(contentStr, sqlEndKeyword, j) {
+						if len(st) == 0 || st[len(st)-1] != sqlBeginKeyword[0] {
+							return nil, fmt.Errorf("error reading contents: unmatched `END` keyword")
+						}
+						st = st[:len(st)-1]
+						j += len(sqlEndKeyword) - 1
+					}
+
+				case sqlSingleQuote, sqlDoubleQuote:
+					quote := contentStr[j]
+					j += 1
+					for j < n && contentStr[j] != quote {
+						j += 1
+					}
+					if j == n {
+						return nil, fmt.Errorf("error reading contents: unmatched quotes")
+					}
+
+				case sqlLineComment[0]:
+					if j+len(sqlLineComment) <= n && contentStr[j:j+len(sqlLineComment)] == sqlLineComment {
+						for j < n && contentStr[j] != '\n' {
+							j += 1
+						}
+					}
+
+				default:
+					// no-op: generic character
+				}
+			}
+
+			if len(st) > 0 {
+				switch st[len(st)-1] {
+				case sqlLeftParenthesis:
+					return nil, fmt.Errorf("error reading contents: unmatched left parenthesis")
+				case sqlBeginKeyword[0]:
+					return nil, fmt.Errorf("error reading contents: unmatched `BEGIN` keyword")
+				default:
+					// should never enter here
+					return nil, fmt.Errorf("error reading contents: unmatched `%c`", st[len(st)-1])
+				}
+			}
+
+			stmt := string(bytes.TrimSpace(content[i:j]))
 			if stmt == "" {
 				continue
 			}
 			result = append(result, stmt)
 		}
-
 	}
 	return result, nil
 }
+
+// matchWord is a simple test to check if it matches the whole word:
+// it checks if the adjacent charactes are not letters if they exist.
+func matchWord(s, word string, pos int) bool {
+	if pos+len(word) > len(s) || s[pos:pos+len(word)] != word {
+		return false
+	}
+	if pos > 0 && unicode.IsLetter(rune(s[pos-1])) {
```

Comment:
What about '7BEGIN'? Is it valid?

Reply:
Not really, I can add this check as well. I wasn't trying to be comprehensive in those checks though. Probably a lot of other edge cases not handled, but it should cover the basic cases, and the rest, the DB will fail the query if it's invalid.